File: Hilos Java/HILOS_METODOS/ReentraLockEjemplo.java
import java.util.concurrent.locks.ReentrantLock;

public class ReentraLockEjemplo {
    private static ReentrantLock lock = new ReentrantLock();

    private static int contador = 0;

    public static void main(String[] args) {
          
        Runnable tarea = () -> {
            for(int i = 0; i<5; i++){
                lock.lock();
                try{
                    contador++;
                    System.out.println(Thread.currentThread().getName() + " incrementó el contador a: "+ contador);

                    try {
                        Thread.sleep(500);
                    } catch (Exception e) {
                        System.out.println("Hilo interrumpido por: "+ e.getMessage());
                    }
                }
                finally{
                    lock.unlock();
                }
            }
        };

        Thread hilo1 = new Thread(tarea, "Hilo-1");
        
        Thread hilo2 = new Thread(tarea, "Hilo-2");

        hilo1.start();

        hilo2.start();
    }
}

File: HILOS_METODOS/ReentraLockEjemplo.py
import threading
import time

class ReentraLockEjemplo:
    lock = threading.RLock()
    contador = 0

    @staticmethod
    def tarea():
        for _ in range(5):
            with ReentraLockEjemplo.lock:
                ReentraLockEjemplo.contador += 1
                print(f"{threading.current_thread().name} incrementó el contador a: {ReentraLockEjemplo.contador}")
                try:
                    time.sleep(0.5)
                except Exception as e:
                    print(f"Hilo interrumpido por: {e}")

if __name__ == "__main__":
    hilo1 = threading.Thread(target=ReentraLockEjemplo.tarea, name="Hilo-1")
    hilo2 = threading.Thread(target=ReentraLockEjemplo.tarea, name="Hilo-2")

    hilo1.start()
    hilo2.start()

    hilo1.join()
    hilo2.join()
